modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/PageMemoryMvPartitionStorage.java

```diff
@@ -0,0 +1,515 @@
+        }
+    }
+
+    private long versionChainLinkFrom(RowId rowId) {
+        if (rowId.partitionId() != partitionId) {
+            throw new IllegalArgumentException("I own partition " + partitionId + " but I was given RowId with partition "
+                    + rowId.partitionId());
+        }
+
+        LinkRowId linkRowId = (LinkRowId) rowId;
+
+        return linkRowId.versionChainLink();
+    }
+
+    @Nullable
+    private ByteBufferRow findLatestRowVersion(VersionChain versionChain, UUID txId, Predicate<BinaryRow> keyFilter) {
+        RowVersion rowVersion = findLatestRowVersion(versionChain);
+        ByteBufferRow row = rowVersionToBinaryRow(rowVersion);
+
+        if (!keyFilter.test(row)) {
+            return null;
+        }
+
+        throwIfChainBelongsToAnotherTx(versionChain, txId);
+
+        return row;
+    }
+
+    private RowVersion findLatestRowVersion(VersionChain versionChain) {
+        return findRowVersion(versionChain.headLink());
+    }
+
+    private RowVersion findRowVersion(long nextRowPartitionlessLink) {
+        long nextLink = PartitionlessLinks.addPartitionIdToPartititionlessLink(nextRowPartitionlessLink, partitionId);
+
+        try {
+            return rowVersionDataPageReader.getRowByLink(nextLink);
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Row version lookup failed");
+        }
+    }
+
+    private void throwIfChainBelongsToAnotherTx(VersionChain versionChain, UUID txId) {
+        if (versionChain.transactionId() != null && !txId.equals(versionChain.transactionId())) {
+            throw new TxIdMismatchException();
+        }
+    }
+
+    @Nullable
+    private ByteBufferRow rowVersionToBinaryRow(RowVersion rowVersion) {
+        if (rowVersion.isTombstone()) {
+            return null;
+        }
+
+        return new ByteBufferRow(rowVersion.value());
+    }
+
+    @Nullable
+    private ByteBufferRow findRowVersionInChain(
+            VersionChain versionChain,
+            @Nullable UUID transactionId,
+            @Nullable Timestamp timestamp,
+            Predicate<BinaryRow> keyFilter
+    ) {
+        assert transactionId != null ^ timestamp != null;
+
+        if (transactionId != null) {
+            return findLatestRowVersion(versionChain, transactionId, keyFilter);
+        } else {
+            ByteBufferRow row = findRowVersionByTimestamp(versionChain, timestamp);
+            return keyFilter.test(row) ? row : null;
+        }
+    }
+
+    @Nullable
+    private ByteBufferRow findRowVersionByTimestamp(VersionChain versionChain, Timestamp timestamp) {
+        long nextRowPartitionlessLink = versionChain.headLink();
+
+        while (true) {
+            RowVersion rowVersion = findRowVersion(nextRowPartitionlessLink);
+            Timestamp rowVersionTs = rowVersion.timestamp();
+            if (rowVersionTs != null && rowVersionTs.beforeOrEquals(timestamp)) {
```

Comment:
I would actually move a comparison method into RowVersion. First reason - it bloats the API of Timestamp. Seconds - it allows you to store timestamp as two longs inside of RowVersion, which is nice. Third - no null-checks and stuff.

Reply:
It's a very natural operation to have on `Timestamp`, it basically pertains to the abstraction (note that it uses the domain-specific 'before' instead of the general 'less'). We bloat an API when we add an operation that is not too cohesive with the type, but here it's not the case. Making a user writing in terms of `compareTo(...) <=> x` produce less readable code. I am sure that we will need to compare timestamps in '<=' sense a lot, so this method seems really useful.

Storing a timestamp as two longs means breaking encapsulation of Timestamp (who knows how will its internals change in the future?). It's an optimization, and at this point it's premature for sure.